VBoxManage

```diff
@@ -6,6 +6,11 @@
 # License: 3-clause BSD-style license (see LICENSE file)
 # Version: 6.1.0
 
+# define the real sed command to use. GNU sed is needed
+_vbmc_sed() {
+  sed "$ @"
```

Comment:
Shouldn't that be 
```bash
sed "$@"
```
so, without the space between `$` and `@`?

Reply:
Also, maybe providing env variable which points to the right sed impementation could be useful, like
```bash
sed_exec=${GNUSED:-sed}
```
than if you like to use different sed for this script, just clutter your environment with GNUSED variable ;)

Reply:
Ah, and finally. Can we enclose this function in completion  in `_VBoxManage`?

Reply:
Sorry for `"$ @"` with the space. Definitively not my day... I add a commit to fix that.

Reply:
+1 for the env var. it's such a quick way to select the needed command. But then do you still think we need the function ? Instead of invoking sed with  `_vbmc_sed`, it's as easy to use `${GNUSED:-sed}`. If you feel both a function and the env var is needed, then that's ok for me, of course. I'll wait for your answer and add another commit according to you choice : fucntion alone, env var alone, or both.

Reply:
Just a question about nesting the functions. As far as I know, bash has no concept of nested function. So this only delays the definition of the nested function until `_VBoxManage` is called for the first time. The drawback is that the functions get redefined each time `_VBoxManage` is called. I've scratched my head trying to find out what you are achieving but could not guess. There must be some subtleties I can't see.

Anyway, I just wait for your answer to the envvar question and move the function inside `_VBoxManage`

Reply:
So the only benefit of having function defined within other function I can think of is that we don't clutter global space with random functions. And since we already have a quite a number of functions defined within `_VBoxManage`, I guess that another one will not become a big game changer.

Moreover, if we agree on using env variable for providing right sed executable, than function will be totally unnecessary, i.e.
```bash
_VBoxManage() {
    …
    SED_EXEC=${GNUSED:-sed}
    …
        hdds=$(VBoxManage list hdds | \
            grep -A 1 'normal (base)' | \
            grep "Location:" | \
            $SED_EXEC 's/Location:\s\+//' | \
            $SED_EXEC 's/\s/\\ /g' | \
            tr '\n' '|' | \
            $SED_EXEC 's/|$//')
    …
```
So it will be even easier for you to set whatever you like as `SED_EXEC` in one place, either by providing env variable or by setting it directly in a script. Unless, of course, you have some other functionality in mind the sed selection function will provide.

Thoughts?

Reply:
I agree with you. having an env var makes the function useless. So let's go for the env var.

About nested functions, I see what the goal is. Unfortunately that doesn't work in bash. Functions are always global to the current shell. Defining a function anywhere defines it globally, so defining function inside other functions clutters the global namespace anyway because bash has only one namespace. Try this. Once you've run `func_with_nested_func`, the ‘external’ `somefunc` is redefined by the ‘internal’ definition.

```shell
somefunc() {
  echo "I'm somefunc, defined at top level"
}
func_with_nested_func() {
  echo "running func_with_nested_func and redefining somefunc"
  somefunc() {
    echo "I'm somefunc, defined inside func_with_nested_func"
  }
  echo "calling somefunc from func_with_nested_func"
  somefunc
}

somefunc
func_with_nested_func
somefunc
declare -f somefunc
```
As stated in SO answer [31316688/1831273](https://stackoverflow.com/a/31316688/1831273), it's possible to get nested private functions if the external function is defined with `()`, so it runs in a subshell. eg

```shell
_VBoxManage() (
  cmp_list() { echo "I'm an invisible from outside _VBoxManage"; }
  ...
)
```
But then that function's COMPREPLY variable would be out of reach of the complete mechanism, making it completely useless...

My next commit proposes to use the gloval var `VBMC_SED=${GNUSED:-sed}`. The var needs be global, local vars aren't available inside `_hdd_comp` and the other funcs.